mountpoint-s3/src/inode.rs

```diff
@@ -141,19 +141,32 @@ impl Superblock {
                 error!("forget called on inode {ino} already removed from the superblock");
             }
             Some(inode) => {
+                // We have to clone the inode so it will not hold reference to the lock which we want to drop as soon as possible.
+                let inode = inode.clone();
+                drop(inodes);
                 logging::record_name(inode.name());
                 let new_lookup_count = inode.dec_lookup_count(n);
                 if new_lookup_count == 0 {
                     // Safe to remove, kernel no longer has a reference to it.
                     trace!(ino, "removing inode from superblock");
-                    let inode = inodes.remove(&ino).expect("we just retrieved it");
-
+                    let inode = self
+                        .inner
+                        .inodes
+                        .write()
+                        .unwrap()
+                        .remove(&ino)
+                        .expect("we just retrieved it");
```

Comment:
well, technically, 2 concurrent `forget` calls on a node with `lookup_count == 1` may result in this failure, since we're not holding the write lock on the collection (but it seems that kernel should never call forget like that, right?)

Reply:
That's true, let's make sure we do not crash here even if there are 2 concurrent forgets.